### PR TITLE
Predictor cache

### PR DESCRIPTION
**Pull request by ea-rus**
Changes:
- Added module for caching
  -  Available cache types: file, redis
- cache predictor results in ApplyPredictorStep 

Docs:
 https://github.com/mindsdb/mindsdb/blob/predictor-cache/mindsdb/utilities/cache.py#L2

#2411

### CODE REVIEW

File: mindsdb/utilities/cache.py [hunk 1]
```diff
@@ -1,121 +1,230 @@
...
+
 class BaseCache(ABC):
-    def __init__(self):
+    def __init__(self, max_size=None, serializer=None):
         self.config = Config()
+        if max_size is None:
+            max_size = self.config["cache"].get("max_size", 50)
+        self.max_size = max_size
+        if serializer is None:
+            self.serializer = self.config["cache"].get('serializer', dill)
```
Comment (StpMax) [inline mindsdb/utilities/cache.py:92]: config is json file - there may be only text value 'dill' in it

**Comment by StpMax** [comment]
Why not to use build in method `df.to_pickle` ? I did few tests, looks like it works at least x2 faster, then dump-to-json.
Also, may be for hash use another method? https://stackoverflow.com/a/62754084

**Review by StpMax — CHANGES_REQUESTED** [review]

File: mindsdb/utilities/cache.py (hunk 1 above)
Comment (ea-rus) [reply mindsdb/utilities/cache.py]: Changed to string

**Comment by ea-rus** [comment]
`Why not to use build in method df.to_pickle`
added set_df and get_df methods to cache dataframes

`https://stackoverflow.com/a/62754084`
pd.util.hash_pandas_object(df) unable to hash list and dict types.

**Review by StpMax — APPROVED** [review]

**Pull request merged**